src/main/java/sx/blah/discord/handle/impl/obj/Webhook.java

```diff
@@ -0,0 +1,166 @@
+package sx.blah.discord.handle.impl.obj;
+
+import org.apache.http.entity.StringEntity;
+import sx.blah.discord.Discord4J;
+import sx.blah.discord.api.IDiscordClient;
+import sx.blah.discord.api.IShard;
+import sx.blah.discord.api.internal.DiscordClientImpl;
+import sx.blah.discord.api.internal.DiscordEndpoints;
+import sx.blah.discord.api.internal.DiscordUtils;
+import sx.blah.discord.api.internal.json.objects.WebhookObject;
+import sx.blah.discord.api.internal.json.requests.WebhookEditRequest;
+import sx.blah.discord.handle.impl.events.WebhookUpdateEvent;
+import sx.blah.discord.handle.obj.*;
+import sx.blah.discord.util.*;
+
+import java.io.UnsupportedEncodingException;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.Optional;
+
+public class Webhook implements IWebhook {
+
+	protected final String id;
+	protected final IDiscordClient client;
+	protected final IChannel channel;
+	protected final IUser author;
+	protected volatile String name;
+	protected volatile String avatar;
+	protected final String token;
+
+	public Webhook(IDiscordClient client, String name, String id, IChannel channel, IUser author, String avatar, String token) {
+		this.client = client;
+		this.name = name;
+		this.id = id;
+		this.channel = channel;
+		this.author = author;
+		this.avatar = avatar;
+		this.token = token;
+	}
+
+	@Override
+	public String getID() {
+		return id;
+	}
+
+	@Override
+	public IDiscordClient getClient() {
+		return client;
+	}
+
+	@Override
+	public IShard getShard() {
+		return channel.getShard();
+	}
+
+	@Override
+	public IWebhook copy() {
+		return new Webhook(client, name, id, channel, author, avatar, token);
+	}
+
+	@Override
+	public IGuild getGuild() {
+		return channel.getGuild();
+	}
+
+	@Override
+	public IChannel getChannel() {
+		return channel;
+	}
+
+	@Override
+	public IUser getAuthor() {
+		return author;
+	}
+
+	@Override
+	public String getDefaultName() {
+		return name;
+	}
+
+	@Override
+	public String getDefaultAvatar() {
+		return avatar;
+	}
+
+	@Override
+	public String getToken() {
+		return token;
+	}
+
+	private void edit(Optional<String> name, Optional<String> avatar) throws MissingPermissionsException, RateLimitException, DiscordException {
```

Comment:
Don't use Optionals here. I know it's used in other places in the library but they will be removed eventually because it's bad practice.